runtime/parachains/src/paras.rs

```diff
@@ -763,6 +771,79 @@ pub mod pallet {
 			Ok(())
 		}
 
+		/// Adds the validation code to the storage.
+		///
+		/// The code will not be added if it is already present. Additionally, if PVF pre-checking
+		/// is running for that code, it will be instantly accepted.
+		///
+		/// Otherwise, the code will be added into the storage. Note that the code will be added
+		/// into storage with reference count 0. This is to account the fact that there are no users
+		/// for this code yet. The caller will have to make sure that this code eventually gets
+		/// used by some parachain or removed from the storage to avoid storage leaks. For the latter
+		/// prefer to use the `poke_unused_validation_code` dispatchable to raw storage manipulation.
+		///
+		/// This function is mainly meant to be used for upgrading parachains that do not follow
+		/// the go-ahead signal while the PVF pre-checking feature is enabled.
+		#[pallet::weight(<T as Config>::WeightInfo::add_trusted_validation_code(validation_code.0.len() as u32))]
+		pub fn add_trusted_validation_code(
+			origin: OriginFor<T>,
+			validation_code: ValidationCode,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			let code_hash = validation_code.hash();
+
+			if let Some(vote) = <Self as Store>::PvfActiveVoteMap::get(&code_hash) {
+				// Remove the existing vote.
+				PvfActiveVoteMap::<T>::remove(&code_hash);
+				PvfActiveVoteList::<T>::mutate(|l| {
+					if let Ok(i) = l.binary_search(&code_hash) {
+						l.remove(i);
+					}
+				});
+
+				let cfg = configuration::Pallet::<T>::config();
+				Self::enact_pvf_accepted(
+					<frame_system::Pallet<T>>::block_number(),
+					&code_hash,
+					&vote.causes,
+					vote.age,
+					&cfg,
+				);
+				return Ok(())
+			}
+
+			if <Self as Store>::CodeByHash::contains_key(&code_hash) {
+				// There is no vote, but the code exists. Nothing to do here.
+				return Ok(())
+			}
+
+			// At this point the code is unknown and there is no PVF pre-checking vote for it, so we
+			// can just add the code into the storage.
+			//
+			// NOTE That we do not use `increase_code_ref` here, because the code is not yet used
+			// by any parachain.
+			<Self as Store>::CodeByHash::insert(code_hash, &validation_code);
```

Comment:
So in this case, we completely ignore the onboarding/upgrade machinery? I guess that's fine/intended, but I would like to understand it better: Like how does the actual upgrade work then for example? Code is already on the relay chain, so what steps must a parachain do now to upgrade in this state?

Reply:
A parachain would attempt an upgrade. When that candidate is enacted, the inclusion module would call [`schedule_code_upgrade`](https://github.com/paritytech/polkadot/blob/3599d0a3cdd5755b0c9d4750060679e75ff30aa7/runtime/parachains/src/paras.rs#L1523) and then control will fall into `kick_off_pvf_check` and this [branch](https://github.com/paritytech/polkadot/blob/3599d0a3cdd5755b0c9d4750060679e75ff30aa7/runtime/parachains/src/paras.rs#L1615) will be taken.

From the parachain's standpoint, it will look like instant PVF approval or equally as good old pre-go-ahead signal upgrade.

Does this make sense?

Reply:
yep it does - thanks! So a parachain would still need to get it's full runtime upgrade through the backing phase, although the code is already on chain. We just bypass the voting, because we trust governance to have the code checked by some other means. :bulb: